fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java

```diff
@@ -1442,6 +1452,64 @@ public static TWorkGroup prepareResourceGroup(ConnectContext connect, ResourceGr
         return resourceGroup;
     }
 
+    private List<Pair<Long, TNetworkAddress>> adaptiveChooseNodes(PlanFragment fragment, Map<Long, ComputeNode> candidates,
+                                                                  Set<TNetworkAddress> childUsedHosts) {
+        List<Pair<Long, TNetworkAddress>> childHosts = Lists.newArrayList();
+        for (Map.Entry<Long, ComputeNode> entry : candidates.entrySet()) {
+            TNetworkAddress address = entry.getValue().getAddress();
+            if (childUsedHosts.contains(address)) {
+                childHosts.add(Pair.create(entry.getKey(), address));
+            }
+        }
+
+        long maxOutputOfRightChild = fragment.getChildren().stream()
+                .sorted(Comparator.comparing(e -> e.getPlanRoot().getId().asInt())).skip(1)
+                .map(e -> e.getPlanRoot().getCardinality()).reduce(Long::max)
+                .orElse(fragment.getChild(0).getPlanRoot().getCardinality());
+        long outputOfMostLeftChild = fragment.getChild(0).getPlanRoot().getCardinality();
+
+
+        long baseNodeNums = Math.max(1, maxOutputOfRightChild / TINY_SCALE_ROWS_LIMIT / fragment.getPipelineDop());
+        double base = Math.max(Math.E, baseNodeNums);
+
+        long amplifyFactor = Math.round(Math.max(1,
+                Math.log(outputOfMostLeftChild / TINY_SCALE_ROWS_LIMIT / fragment.getPipelineDop()) / Math.log(base)));
+
+        long nodeNums = Math.min(amplifyFactor * baseNodeNums, candidates.size());
```

Comment:
what is the meaning of `amplifyFactor * baseNodeNums` ? why?

Reply:
The current strategy is relative simple. Use the child with the most output rows among the right children to determine the basic number of instances, and then adjust the number of instances appropriately based on the relationship between the output rows of the left child and the output rows of the right child.

Reply:
what if data Explosion like after join?

Reply:
I think the following two points may help alleviate the data explosion problem
1. We select the base node based on the largest amount of data in the right child, which helps avoid selecting a few nodes under large amounts of data in most scenarios.
2. If the input amount is realtive small, but data explosion occurs due to join, it is more likely that the join is  1:N scenarios which means data skew may occurre, and selecting a large number nodes may not alleviate this situation.